tools/crm_mon.c

```diff
@@ -2994,60 +2994,41 @@ print_failed_actions(FILE *stream, pe_working_set_t *data_set)
 static stonith_history_t *
 reduce_stonith_history(stonith_history_t *history)
 {
-    stonith_history_t *new = NULL, *hp, *np, *tmp;
+    stonith_history_t *new = NULL, *hp, *np;
 
     for (hp = history; hp; ) {
-        for (np = new; np; np = np->next) {
-            if ((hp->state == st_done) || (hp->state == st_failed)) {
-                /* action not in progress */
-                if (safe_str_eq(hp->target, np->target) &&
-                    safe_str_eq(hp->action, np->action) &&
-                    (hp->state == np->state)) {
-                    if ((hp->state == st_done) ||
-                        safe_str_eq(hp->delegate, np->delegate)) {
-                        /* replace or purge */
-                        if (hp->completed < np->completed) {
+        stonith_history_t *hp_next = hp->next;
+
+        hp->next = NULL;
+        if (!new) {
+            new = hp;
+        } else {
+            for (np = new; ; np = np->next) {
+                if ((hp->state == st_done) || (hp->state == st_failed)) {
+                    /* action not in progress */
+                    if (safe_str_eq(hp->target, np->target) &&
+                        safe_str_eq(hp->action, np->action) &&
+                        (hp->state == np->state) &&
+                        ((hp->state == st_done) ||
+                         safe_str_eq(hp->delegate, np->delegate)) &&
+                        (hp->completed < np->completed)) {
                             /* purge older hp */
-                            tmp = hp->next;
-                            hp->next = NULL;
                             stonith_history_free(hp);
-                            hp = tmp;
                             break;
```

Comment:
At this point, since hp is the last item in new, the item before it will have its next (still pointing to hp) become invalid.

Reply:
Does it mean that the NULL set is missing?
Certainly, if you don't set NULL, problems will occur with subsequent sets of next.

```
(snip)
                            /* purge older hp */
                            stonith_history_free(hp);
                            hp = NULL;
                            break;
                    }
                }
                if (!np->next) {
                    np->next = hp;
                    break;

```

Reply:
Sorry, I misread the code. There is no problem with the next pointer.

One point I'm still unclear about: Isn't (hp->completed < np->completed) unnecessary? The list was sorted with newest first, so that should always be true.